org.eclipse.ui.console/src/org/eclipse/ui/console/ConsolePlugin.java

```diff
@@ -193,5 +211,31 @@ public void stop(BundleContext context) throws Exception {
 		super.stop(context);
 	}
 
+	/**
+	 * @since 3.12
+	 */
+	public void addViewer(StyledText viewer, IConsolePageParticipant participant) {
+		viewers.put(viewer, participant);
+		if (checkPerformanceOnce) {
+			AnsiConsoleUtils.checkConsolePerformanceSettings();
+			checkPerformanceOnce = false;
+		}
+	}
 
+	/**
+	 * @since 3.12
+	 */
+	public void removeViewerWithPageParticipant(IConsolePageParticipant participant) {
```

Comment:
Does it have to be a public API? Couldn't it be internal?

Reply:
`removeViewerWithPageParticipant` is called from `AnsiConsolePageParticipant.init`
And `addViewer` is called from `AnsiConsolePageParticipant.dispose`

They add / remove a `IConsolePageParticipant` to 
`private final Map<StyledText, IConsolePageParticipant> viewers`
Which I've also added to `ConsolePlugin`

I think that everything (map + the methods to add / remove) can be moved to `AnsiConsolePageParticipant`
But I am not sure about the lifespan of `AnsiConsolePageParticipant`.
Maybe make the `viewers` map static? A singleton?
The plugin I knew it can "survive" as long as it is needed.

I welcome any advice.

Reply:
Or, is there a way to tag a public method (from Java perspective) as internal for Eclipse?
Some annotation?

Reply:
> I think that everything (map + the methods to add / remove) can be moved to AnsiConsolePageParticipant

Let's try that.

> Maybe make the viewers map static? A singleton?

Yes, keeping a static map should be good enough.

> Or, is there a way to tag a public method (from Java perspective) as internal for Eclipse?

There is @noreference in the javadoc but it's not really good to use in general. So moving code to an internal class is greatly preferred.

Reply:
Hmm, it looks like `@noreference` does the job.

I've added `@noreference` to `addViewer` and `removeViewerWithPageParticipant`, removed `@since`,
and reverted the `Bundle-Version` in `MANIFEST.MF` to `3.11.300.qualifier` (as it was before I touched it)

But moving the two methods to `AnsiConsolePageParticipant` would be cleaner, if that is safe to do.

Reply:
Done.

Reply:
As mentioned, @reference is to be avoided. Please move to an internal class instead.

Reply:
Done.